### PR TITLE
feat: New istio config format

### LINKED ISSUE #1247: New configuration format
This Issue comes from this design document: https://docs.google.com/document/d/12X1-9nhjAhpf-Wlt3RbdoMbvx31zFBva/edit#heading=h.gjdgxs

The proposal is to make the configuration more flexible, more extensible. This issue focus only on the new configuration format, without any filtering capabilities (that will come in another issue)

See the design document for more information

### PR DESCRIPTION
**Pull request by pastequo**
# Changes

- Introduce new configuration format

/kind enhancement

Fixes #1247 

**Release Note**

```release-note
A new configuration format has been introduced. Old format is still supported. See https://docs.google.com/document/d/12X1-9nhjAhpf-Wlt3RbdoMbvx31zFBva/edit for more details
```

**Docs**

```docs

```

### CODE REVIEW

File: pkg/reconciler/ingress/config/istio.go [hunk 1]
```diff
@@ -89,17 +116,120 @@ type Istio struct {
...
+	gateways, err := parseNewFormatGateways(gatewaysStr)
+	if err != nil {
+		return ret, fmt.Errorf("failed to parse gateways")
+	}
+
+	ret.IngressGateways = gateways
+
+	localGateways, err := parseNewFormatGateways(localGatewaysStr)
+	if err != nil {
+		return ret, fmt.Errorf("failed to parse local gateways")
```
Comment (dprotaso) [inline pkg/reconciler/ingress/config/istio.go:180]:
```suggestion
		return ret, fmt.Errorf("failed to parse %q: %w", localGatewaysKey, err)
```

File: pkg/reconciler/ingress/config/istio.go [hunk 2]
```diff
@@ -89,17 +116,120 @@ type Istio struct {
...
+	}
+
+	// New format is not defined
+	if !hasGateway && !hasLocalGateway {
+		return ret, nil
+	}
+
+	gateways, err := parseNewFormatGateways(gatewaysStr)
+	if err != nil {
+		return ret, fmt.Errorf("failed to parse gateways")
```
Comment (dprotaso) [inline pkg/reconciler/ingress/config/istio.go:173]:
```suggestion
		return ret, fmt.Errorf("failed to parse %q: %w", externalGatewaysKey, err)
```

File: pkg/reconciler/ingress/config/istio.go [hunk 3]
```diff
@@ -89,17 +116,120 @@ type Istio struct {
...
+
+	ret.IngressGateways = gateways
+
+	localGateways, err := parseNewFormatGateways(localGatewaysStr)
+	if err != nil {
+		return ret, fmt.Errorf("failed to parse local gateways")
+	}
+
+	if len(localGateways) > 1 {
+		return ret, fmt.Errorf("only one local gateways can be defined")
```
Comment (dprotaso) [inline pkg/reconciler/ingress/config/istio.go:184]: Maybe make the error message more clear my mentioning the key that has the problematic value

File: pkg/reconciler/ingress/config/istio.go [hunk 4]
```diff
@@ -89,17 +116,120 @@ type Istio struct {
...
+func parseNewFormat(configMap *corev1.ConfigMap) (*Istio, error) {
+	ret := &Istio{}
+
+	gatewaysStr, hasGateway := configMap.Data[gatewayKey]
+	localGatewaysStr, hasLocalGateway := configMap.Data[localGatewayKey]
+
+	// New format is partially defined
+	if (!hasGateway && hasLocalGateway) || (hasGateway && !hasLocalGateway) {
+		return nil, fmt.Errorf("config is not fully defined")
+	}
```
[the remark this answers is not in the document]
Comment (pastequo) [reply pkg/reconciler/ingress/config/istio.go]: So I implemented the following

- If both format are defined (even partially), the code will error
- default values are used either by the new or old format

I let you resolve the thread if this new version is ok for you

**Pull request merged**